### PR TITLE
history, build: add pre-allowed variables e.g `proxy vars` to history if explicitly specified

### PR DESCRIPTION
**Pull request by flouthoc**
`Buildkit/Docker` only adds pre-allowed variables for e.g `proxy
variables` to history if user explicitly specifies them in
Dockerfile/Containerfile using `ARG`.

By default variables in allowlist e.g `proxy variables` and values will be used
normally but will not be leaked into `docker/OCI` history of images.

A test for following behavior is added with this commit and similar test
can be verified against `Docker/Buildkit`

Closes: https://github.com/containers/buildah/issues/2937

### CODE REVIEW

File: imagebuildah/stage_executor.go [hunk 1]
```diff
@@ -64,6 +65,7 @@ type StageExecutor struct {
 	output          string
 	containerIDs    []string
 	stage           *imagebuilder.Stage
+	processedArgs   []string // contains args which are processed and were added implicitly in Containerfile using ARG
```
Comment (vrothberg) [inline imagebuildah/stage_executor.go:68]:
```suggestion
	argsFromContainerfile   []string
```
?

File: imagebuildah/stage_executor.go [hunk 2]
```diff
@@ -1271,7 +1278,31 @@ func (s *StageExecutor) getBuildArgsResolvedForRun() string {
 			if inImage {
 				envs = append(envs, fmt.Sprintf("%s=%s", key, configuredEnvs[key]))
 			} else {
-				envs = append(envs, fmt.Sprintf("%s=%s", key, value))
+				// by default everything must be added to history
+				// this is configured to false only for special cases
```
Comment (vrothberg) [inline imagebuildah/stage_executor.go:1282]: Please use punctuation in sentences and avoid lower-case only printing; I have a hard time brain-parsing the comments.

File: imagebuildah/stage_executor.go [hunk 3]
```diff
@@ -1271,7 +1278,31 @@ func (s *StageExecutor) getBuildArgsResolvedForRun() string {
...
+				addToHistory := true
+
+				// following value is being assigned from build-args
+				// check if this key belongs to any of the predefined whitelisted args e.g Proxy Variables
+				// and if that arg is not manually set in Containerfile/Dockerfile
+				// then don't write its value to history.
+				// following behaviour ensures parity with docker/buildkit behviour
```
Comment (vrothberg) [inline imagebuildah/stage_executor.go:1289]:
```suggestion
				// following behaviour ensures parity with docker/buildkit behaviour
```

File: imagebuildah/stage_executor.go [hunk 4]
```diff
@@ -1271,7 +1278,31 @@ func (s *StageExecutor) getBuildArgsResolvedForRun() string {
...
+				for _, variable := range config.ProxyEnv {
+					if key == variable {
+						// found in predefined args
+						// so don't add to history
+						// unless user did implicit `ARG <some-predefined-proxy-variable>`
```
Comment (vrothberg) [inline imagebuildah/stage_executor.go:1294]:
```suggestion
						// unless user did explicit `ARG <some-predefined-proxy-variable>`
```

**Review by vrothberg — COMMENTED** [review]
Code LGTM, just a couple of nits.

Can you do a `s/whitelist/allowlist/`? "Whitelisting" is an old non-inclusive term that we want to avoid.

Can you add `Fixes: https://github.com/containers/buildah/issues/2937` to the commit message? That would add some useful information to the git history.

File: imagebuildah/stage_executor.go (hunk 2 above)
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: Modified comments a bit.

**Comment by flouthoc** [comment]
@vrothberg Thanks tweaked comments as requested.

**Review by vrothberg — COMMENTED** [review]
Thank you, LGTM

**Comment by nalind** [comment]
> Can you do a `s/whitelist/allowlist/`? "Whitelisting" is an old non-inclusive term that we want to avoid.

Test function names, too.

**Pull request merged**